src/main/antora/modules/security/pages/security.adoc

```diff
@@ -776,8 +776,352 @@ public class RestFormAuthCustomStoreIT extends ITBase {
+----
+
+The resource that we defined above required only two roles to access it (`user` and `caller2`), but our custom identity store also returned `caller1`. The resource we created tests for this, and as it appears, we indeed had this role.
+
+[NOTE]
+If we hadn't declared `caller1` in `web.xml` (or via an annotation), the test for `caller1` might have returned false. This is however server dependent.
+
+
+
+== Securing an endpoint with a custom authentication mechanism and a custom identity store
+
+In the following example, we'll be securing a REST endpoint using a custom authentication mechanism. A custom authentication mechanism is one we provide ourselves, instead of using one provided by Jakarta Security (such as the Basic HTTP authentication mechanism).
+
+You'll learn how to:
+
+. Define security constraints
+. Define (and implicitly set) a custom authentication mechanism
+. Define (and implicitly set) a custom identity store
+. Use the Jakarta Security `SecurityContext`
+
+=== Write the application
+
+We'll use the same resource and same security constraints as we used for the <<Securing an endpoint with Basic authentication>> example.
+
+==== Define the authentication mechanism
+
+Let's now define a simple authentication mechanism that the security system can use to interact with the caller who tries to access a resource:
+
+[source,java]
+----
+@ApplicationScoped
+public class CustomAuthenticationMechanism implements HttpAuthenticationMechanism {
+    
+    @Inject
+    private IdentityStoreHandler identityStoreHandler;
+
+    @Override
+    public AuthenticationStatus validateRequest(
+        HttpServletRequest request, 
+        HttpServletResponse response, 
+        HttpMessageContext httpMessageContext) throws AuthenticationException {
+        
+        String callerName = request.getHeader("callername");
+        String password = request.getParameter("callerpassword");
+        
+        if (callerName == null || password == null) {
+            return httpMessageContext.doNothing();
+        }
+
+        CredentialValidationResult result = identityStoreHandler.validate(
+            new UsernamePasswordCredential(callerName, password));
+        
+        if (result.getStatus() != VALID) {
+            return httpMessageContext.responseUnauthorized();
+        }
+
+        return httpMessageContext.notifyContainerAboutLogin(
+            result.getCallerPrincipal(), 
+            result.getCallerGroups());
+    }
+    
+}
+----
+
+This custom authentication mechanism interacts with the caller by grabbing two headers from the request: `callername` and `callerpassword`. In case any of them are null, we return a special status; the "do nothing" status. This means there has been no request or attempt to do authentication. If the resource the caller is trying to access is not protected, the caller can access it anonymously. If it is proteced, the caller will not be able to access it.
```

Comment:
You could use callouts here as well, but seems less essential.

Reply:
References to items in the code are absolutely great, but it looks like it's rendered only as a somewhat cramped list beneath the code, instead of somewhat more free flowing sections.

Reply:
I've now tried this style:

<img width="624" alt="example" src="https://github.com/virtua-tech/jakartaee-tutorial/assets/3037006/996ddd8d-047a-42fd-9f9d-2ea408204897">